### DISCUSSION TITLE
Home direction offset inaccuracy

### DISCUSSION THREAD

**Discussion by Sibre3** [iNavFlight/inav#9050] [Q&A]
I'm trying to setup 'Pan servo home direction offset' on my plane without success.
I now understand how `osd_pan_servo_pwm2centideg` should be set.
But my problem is the way this command is tied to a fixed 1000 – 2000uS limit. 
This prevents the Outputs or Mixer rates from limiting the servo travel, without an error being introduced into the Home arrow or the `osd_pan_servo_indicator_show_degrees` value.

In my case the problem is occurring because of the pan servo I am using. It's a GWS S125-1T 2BB sail winch servo. These servos give very smooth 360 degree movement. But it actually rotates approximately 400 degrees between 1000 – 2000uS, with `osd_pan_servo_pwm2centideg` limit being 360 degrees. And there lies my problem.

Is there anyway I can reduce the pan servo travel limit through INAV. Without it directly altering the accuracy of the Home arrow and Pan travel readout?

Thanks in advance.

**Comment by Jetrell** [comment]
As you've noticed. The micro second position of the servo is the reference for the **home arrow** and the **pan degrees of travel**... So that value can't be reduced or increased without it effecting the accuracy of either the _home arrow_, _cardinal marker_ or the _pan travel degrees_.
This should be correctable by setting `osd_pan_servo_pwm2centideg` to value equal to the pan servo's travel degrees.. But I think `36` is the limit.
And I also think the OUTPUTS and RATES are tied to the pan offset..  

P.S. Another thing I thought of...  The pan servo most likely can't be trimmed for equal left/ right travel, without it effecting accuracy ? 
e.g. If the servo's pot provides more travel in one direction than the other.. Which often occurs with most servo's.

@MrD-RC maybe able to confirm these questions ?

**Reply by MrD-RC** [reply]
From what I remember, it takes the pan input as 1000us to 2000us, and the servo half rotation to what you set in ** osd_pan_servo_pwm2centideg** To offset. The max is 180 degrees (360 degree servo rotation). It then calculates the offset by dividing the input with the rotation. 

With regards to the unequal trim. INAV doesn’t take this in to consideration. Historically I tended to set that up on the transmitter. But it would be better if centre, min, and max were set up on the INAV outputs tab; with equal movement each way from centre. For example: min 1150 | centre 1650 | max 2150. Then there would be no offset to worry about, as it will already be handled. Just have the transmitter 1000-2000 with 1500 centre.

**Reply by Sibre3** [reply]
@MrD-RC  Does this indicate that a 500uS travel correlates to 8 home arrow character positions within 180° of camera travel. Based on the premise, the servo PWM value must always remain a constant -500 +500uS. 

I heard of a Youtube content provider, who has experienced the same trouble as myself, with a servo that provides +400° of total rotation, from a 1 turn servo. 
If we try to limit the servo travel to provide 360° of camera movement; either in the transmitter or INAV. To a value of less than 500uS per side. We are left with the home arrow and pan degrees readout calculations being compressed within the reduced travel limits.
This interprets as, 90° of camera travel, reading as 79°. Or 180° of camera travel, reading as 159°. 

The only way I see around this impasse. Would be to provide a higher degree of limit for `osd_pan_servo_pwm2centideg`. Example 41, instead of 36. Or even higher for other 1 turn servos. Of which there absolute travel can not be limited internally, based on gearing.

I am aware that you implemented the code for this feature. 
I would like to thank you for your efforts in doing so. 
For cases like this one, may I request that you would please be willing to increase the maximum limit of `osd_pan_servo_pwm2centideg`. So this great feature can be used to its fullest, by us all.

Thank you

**Reply by MrD-RC** [reply]
That really doesn’t make sense to increase the servo offset by that much. The current maximum (180 degrees) is enough for a 360 degree servos. With the FOV of our FPV cameras, even 270 degrees total servo movement will get you looking back behind you. 

400 degrees of servo travel, I would argue, is an edge case. Most people use a 180 degree servo and extend the limits to get around 270 degrees of movement. Which is plenty. Also, most 360 degree servos that we see are not self centring. Those types of servos can’t be supported. 

Whatever servo you use for the pan, you will need to tune the **osd_pan_servo_pwm2centideg**. If it’s saying 90 degrees of travel is 79 degrees. You need to change the offset until the angle matches. Likewise. If it rotates the wrong way. You would need to make the value a negative.

**Comment by Sibre3** [comment]
The servo is a GWS S125 1T 2BB. https://www.aliexpress.com/item/32482605281.html
Many FPV pilots have use them for camera or antenna panning, for over 7 years. Because of their ONE turn 4096 bit resolution PWM board, with 2 ball bearings on the output.
This is helpful in handling the added weight of a HD camera being attached to the pan mechanism.
It was bonafidepirate who first put me onto this servo.

This servo is a 1 turn proportionally controlled servo. But has slightly more than 360° of mechanical travel.
The problem comes when 1000uS of travel actually produces 400° of mechanical servo travel on most of these 1 turn servo's.
This will provide 200° of travel left or right of center. Or 20° more travel past rear center, in either direction. This is useful in flight when being followed by other planes. 
But the pan offset math can not take this into account with its 36 limit. It doesn't know 1000uS of PWM is actually traveling 200°. Because the maximum INAV software limit for a 1 turn servo is 180° left and right of center.

The problem in this case is the offset limit of 36. 
You stated that I need to increase the offset. As when a 180° travel servo is expanded to 270°.
This will work for a servo with a combine travel less than 360°. Because the expanded offset value is still less than 180° per side. In your example it would go from 18 to 27 total travel offset.
But the 1T servos require 36 to expand to 41.

**Reply by Jetrell** [reply]
Okay.. I think I see what your saying.
When a servo with a mechanical output of 400° of travel, based on a 1000 - 2000uS input control.. With `osd_pan_servo_pwm2centideg = 36`.
The camera will actually rotate to 200° from front center. Back past rear center(180°).
But there is no way for INAV to know that the camera is viewing an angle 200° (red) from front center... From INAV's perspective, it thinks the camera is facing 180° (purple) rearward. But its not. Its facing 200°.




![pie](https://github.com/iNavFlight/inav/assets/47995726/56c52940-bd4a-41b6-b116-64f60fc84bfc)


So the `osd_pan_servo_indicator_show_degrees` will display a lesser numerical value than 180°, when the camera is actually facing 180° rearward (purple). With the home arrow being 1 position less as well.
And that would scale down I imagine. When the camera is actually facing 90°. It would be reading 79° as you said.

**Reply by Sibre3** [reply]
That explains it perfectly.

**Comment by Sibre3** [comment]
@MrD-RC  Thank you for fixing this problem with https://github.com/iNavFlight/inav/pull/11072
